File: packages/shared-types/src/events.ts
import { z } from 'zod';
import { CanonicalContentSchema } from './canonical.js';
import { CapabilityIdSchema } from './capabilities.js';
import { ChannelSchema } from './identity.js';
import { OrderStatusSchema } from './order.js';
import { DaemonProposalSchema } from './daemon.js';

// Every event carries this base envelope. Producers fill it; consumers validate.
// Optional fields are *also* nullable so Python publishers (Pydantic) can serialise
// `None` → `null` without tripping Zod validation on the consumer side.
const EventBaseSchema = z.object({
  event_id: z.string().uuid(),
  occurred_at: z.string().datetime(),
  trace_id: z.string().nullable().optional(),
  span_id: z.string().nullable().optional(),
});

// ── messages.inbound ─────────────────────────────────────────────────────
export const MessagesInboundEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid().nullable(),
  thread_id: z.string().uuid().nullable(),
  principal_id: z.string().uuid(),
  channel: ChannelSchema,
  channel_message_id: z.string(),
  sender_identifier: z.string(),
  recipient_identifier: z.string(),
  content: CanonicalContentSchema,
  raw_payload: z.unknown().nullable().optional(),
});
export type MessagesInboundEvent = z.infer<typeof MessagesInboundEventSchema>;

// ── messages.outbound ────────────────────────────────────────────────────
export const MessagesOutboundEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid().nullable(),
  thread_id: z.string().uuid().nullable(),
  target_channel: ChannelSchema,
  target_identifier: z.string(),
  source_phone_number_id: z.string().nullable().optional(),
  content: CanonicalContentSchema,
  template_name: z.string().nullable().optional(),
  template_variables: z.record(z.string()).nullable().optional(),
  requires_window_check: z.boolean().default(true),
  priority: z.enum(['low', 'normal', 'high']).default('normal'),
});
export type MessagesOutboundEvent = z.infer<typeof MessagesOutboundEventSchema>;

// ── agent.actions ────────────────────────────────────────────────────────
export const AgentActionEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid().nullable(),
  thread_id: z.string().uuid().nullable(),
  action_type: z.enum([
    'sub_agent_dispatched',
    'tool_called',
    'escalation_triggered',
    'session_started',
    'session_ended',
    'confidence_assessed',
  ]),
  sub_agent: z.string().optional(),
  tool: CapabilityIdSchema.optional(),
  tool_input: z.unknown().optional(),
  tool_output: z.unknown().optional(),
  duration_ms: z.number().int().nonnegative().default(0),
  model_used: z.string().optional(),
  provider: z.string().optional(),
  tokens: z
    .object({
      input: z.number().int().nonnegative().default(0),
      output: z.number().int().nonnegative().default(0),
      cached: z.number().int().nonnegative().default(0),
    })
    .default({ input: 0, output: 0, cached: 0 }),
  cost_paise: z.number().int().nonnegative().default(0),
});
export type AgentActionEvent = z.infer<typeof AgentActionEventSchema>;

// ── blueprint.mutations ──────────────────────────────────────────────────
export const BlueprintMutationEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid(),
  version_from: z.number().int().nonnegative(),
  version_to: z.number().int().positive(),
  mutated_by_principal: z.string().uuid().nullable(),
  mutation_source: z.enum(['veda', 'dashboard', 'api', 'migration']),
  diff: z.record(z.unknown()).default({}),
  full_blueprint: z.record(z.unknown()),
});
export type BlueprintMutationEvent = z.infer<typeof BlueprintMutationEventSchema>;

// ── orders ───────────────────────────────────────────────────────────────
export const OrderEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid(),
  order_id: z.string().uuid(),
  order_number: z.string(),
  transition: z.string(), // e.g. "created->confirmed"
  to_status: OrderStatusSchema,
  total_paise: z.number().int().nonnegative(),
  principal_id: z.string().uuid(),
  line_items: z.array(z.record(z.unknown())).default([]),
});
export type OrderEvent = z.infer<typeof OrderEventSchema>;

// ── escalations ──────────────────────────────────────────────────────────
export const EscalationEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid(),
  thread_id: z.string().uuid(),
  reason: z.string(),
  escalate_to: z.enum(['owner', 'admin', 'operator']),
  assigned_to: z.string().uuid().nullable().optional(),
});
export type EscalationEvent = z.infer<typeof EscalationEventSchema>;

// ── daemon.proposals ─────────────────────────────────────────────────────
export const DaemonProposalEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid(),
  proposal: DaemonProposalSchema,
});
export type DaemonProposalEvent = z.infer<typeof DaemonProposalEventSchema>;

// ── billing.usage ────────────────────────────────────────────────────────
export const BillingUsageEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid().nullable(),
  provider: z.enum(['anthropic', 'azure_openai', 'sarvam', 'azure_speech']),
  model: z.string(),
  task_type: z.string(),
  input_tokens: z.number().int().nonnegative().default(0),
  output_tokens: z.number().int().nonnegative().default(0),
  cached_tokens: z.number().int().nonnegative().default(0),
  cost_paise: z.number().int().nonnegative(),
});
export type BillingUsageEvent = z.infer<typeof BillingUsageEventSchema>;

// ── audit.log ────────────────────────────────────────────────────────────
export const AuditEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid().nullable(),
  principal_id: z.string().uuid().nullable(),
  event_type: z.string(),
  entity_type: z.string().optional(),
  entity_id: z.string().optional(),
  payload: z.record(z.unknown()).default({}),
  ip_address: z.string().optional(),
  user_agent: z.string().optional(),
});
export type AuditEvent = z.infer<typeof AuditEventSchema>;

// ── integrations.events ──────────────────────────────────────────────────
export const IntegrationEventSchema = EventBaseSchema.extend({
  tenant_id: z.string().uuid(),
  integration: z.string(),
  event_type: z.string(),
  payload: z.record(z.unknown()).default({}),
});
export type IntegrationEvent = z.infer<typeof IntegrationEventSchema>;

File: packages/python-shared/veda_shared/schemas/events.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from veda_shared.schemas.canonical import CanonicalContent
from veda_shared.schemas.capabilities import CapabilityId
from veda_shared.schemas.identity import Channel


class _EventBase(BaseModel):
    model_config = ConfigDict(extra="ignore")

    event_id: str
    occurred_at: datetime
    trace_id: str | None = None
    span_id: str | None = None


class MessagesInboundEvent(_EventBase):
    tenant_id: str | None = None
    thread_id: str | None = None
    principal_id: str
    channel: Channel
    channel_message_id: str
    sender_identifier: str
    recipient_identifier: str
    content: CanonicalContent
    raw_payload: Any | None = None


class MessagesOutboundEvent(_EventBase):
    tenant_id: str | None = None
    thread_id: str | None = None
    target_channel: Channel
    target_identifier: str
    source_phone_number_id: str | None = None
    content: CanonicalContent
    template_name: str | None = None
    template_variables: dict[str, str] | None = None
    requires_window_check: bool = True
    priority: Literal["low", "normal", "high"] = "normal"


class _Tokens(BaseModel):
    input: int = 0
    output: int = 0
    cached: int = 0


class AgentActionEvent(_EventBase):
    tenant_id: str | None = None
    thread_id: str | None = None
    action_type: Literal[
        "sub_agent_dispatched",
        "tool_called",
        "escalation_triggered",
        "session_started",
        "session_ended",
        "confidence_assessed",
    ]
    sub_agent: str | None = None
    tool: CapabilityId | None = None
    tool_input: Any | None = None
    tool_output: Any | None = None
    duration_ms: int = 0
    model_used: str | None = None
    provider: str | None = None
    tokens: _Tokens = Field(default_factory=_Tokens)
    cost_paise: int = 0


class BlueprintMutationEvent(_EventBase):
    tenant_id: str
    version_from: int
    version_to: int
    mutated_by_principal: str | None = None
    mutation_source: Literal["veda", "dashboard", "api", "migration"]
    diff: dict[str, Any] = Field(default_factory=dict)
    full_blueprint: dict[str, Any]


class OrderEvent(_EventBase):
    tenant_id: str
    order_id: str
    order_number: str
    transition: str
    to_status: str
    total_paise: int
    principal_id: str
    line_items: list[dict[str, Any]] = Field(default_factory=list)


class EscalationEvent(_EventBase):
    tenant_id: str
    thread_id: str
    reason: str
    escalate_to: Literal["owner", "admin", "operator"]
    assigned_to: str | None = None


class DaemonProposalEvent(_EventBase):
    tenant_id: str
    proposal: dict[str, Any]


class BillingUsageEvent(_EventBase):
    tenant_id: str | None = None
    provider: Literal["anthropic", "azure_openai", "sarvam", "azure_speech"]
    model: str
    task_type: str
    input_tokens: int = 0
    output_tokens: int = 0
    cached_tokens: int = 0
    cost_paise: int


class AuditEvent(_EventBase):
    tenant_id: str | None = None
    principal_id: str | None = None
    event_type: str
    entity_type: str | None = None
    entity_id: str | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    ip_address: str | None = None
    user_agent: str | None = None


class IntegrationEvent(_EventBase):
    tenant_id: str
    integration: str
    event_type: str
    payload: dict[str, Any] = Field(default_factory=dict)
